// ocap/event.js
var Config  = require( 'config/event.js' ).Config;

exports.Event = Event;

function Event( net, finishHandler ) {

  var self = this;

  self.__finishHandler = finishHandler;

  self.getDetails = function ( connectionId, connection ) {
    var cId      = connectionId.split( '@' );
    var instance = cId[1];
    if ( !connection || !( connection.sessionCookie ) ) {
      self.__finishHandler( connectionId );
      return;
    }
    var userType;
    if ( connection.user_type == Config.BUYER ) {
      userType = 'buyer';
    }
    else if ( connection.user_type = Config.SUPPLIER ) {
      userType = 'supplier';
    }
    self.net.get( 'events/' + userType + '/' + connection.org_id,
      function ( error ) {
        if ( !error )
          error = Config.networkError;
        connection.lastAction = Config.failedSignInMsg;
        console.error( Config.errorPrefix + ' ' + error );
        self.__finishHandler( connectionId, error );
      },
      function ( response ) {
        connection.lastAction = Config.eventsDetailAction;
        console.log( instance   + ' - ' + Config.eventsDetailMsg );
        var payload           = JSON.parse( response.body ).payload;
        connection.lastActive = new Date( payload.created );
        var events            = payload.response.events; 
        for ( e in events ) {
          var theEvent = events[e];
          var eventId  = theEvent.event_id;
          connection.events = connection.events || {};
          connection.events[eventId] = {};
          var event = connection.events[eventId]
          console.log(instance + ' - ' + Config.foundEventMsg + eventId );
          for ( key in theEvent ) {
            if ( /_time$/.test( key ) )
              event[key] = new Date( theEvent[key] );
            else 
              event[key] = theEvent[key];
          }
          var isLive       = event.is_live;
          var isBuyerLive  = event.is_buyer_live;
          var isActive     = event.is_active;
          var startTime    = event.start_time;
          var endTime      = event.end_time;
          var isEncumbered = event.is_encumbered;
          var newCheckDate = Util.iso8601Date( event.new_check_date );
          var payThruDate  = Util.iso8601Date( event.pay_thru_date );
          var startsIn     = Util.timeElapse( ( new Date() ).getTime(), startTime.getTime() );
          var duration     = Util.timeElapse( startTime.getTime(), endTime.getTime() );
          var eventStatus = ''
          if ( isActive && isLive && isBuyerLive ) {
            eventStatus = Config.liveStatus;
          }
          else if ( isActive && isBuyerLive ) {
            eventStatus = Config.awaitingStatus;
          }
          else if ( isActive ) {
            eventStatus = Config.preofferStatus;
          }
          console.log( instance + ' - #' + eventId + ': ' + Config.countdownMsg + ': ' + startsIn );
          console.log( instance + ' - #' + eventId + ': ' + Config.durationMsg  + ': ' + duration );
          console.log( instance + ' - #' + eventId + ': ' + Config.newCheckMsg  + ': ' + newCheckDate );
          console.log( instance + ' - #' + eventId + ': ' + Config.payThruMsg   + ': ' + payThruDate );
          console.log( instance + ' - #' + eventId + ': ' + Config.statusPrefix + ': ' + eventStatus );
        }
        self.__finishHandler( connectionId );
      },
      instance,
      connection.sessionCookie
    );
  };

  self.isDebugging = false;

  self.lockBasket = function ( connectionId, connection, eventId, supplierId, basketId, statusCode ) {

    var lockBasketCallback = function ( connectionId, error ) {
      if ( error ) {
        console.log( Config.errorFlip + ': ' + connectionId );
        self.__postCommandCallback( connectionId );
        return;
      }
      else {
        console.log( Config.successFlip );
      }
    };

    if ( connection ) { 
      if ( connection.user_type == Config.BUYER ) {
        var cId       = connectionId.split( '@' );
        var instance  = cId[1];
        var emulating = cId[2];
        var payload = {
          details: [ supplierId, 'basket', basketId, 'status' ],
          supplierId : supplierId,
          status     : statusCode
        }
        self.net.post( payload, 'events/' + eventId + '/' + supplierId + '/basket/' + basketId + '/status',
          function ( error ) {
            if ( !error )
              error = Config.networkError;
            connection.lastAction = Config.failedSignInMsg;
            console.error( Config.errorPrefix + ' ' + error );
            self.__finishHandler( connectionId, error );
          },
          function ( response ) {
            connection.lastAction    = Config.userDetailsAction;
            console.log( instance + ' - ' + Config.userDetailsMsg );
            var payload              = JSON.parse( response.body ).payload;
            connection.lastActive    = new Date( payload.created );
            delete payload.created;
          },
          instance,
          connection.sessionCookie
        );
      }
      else if ( connection.user_type = Config.SUPPLIER ) {
        console.log( Config.errorSupplierCant );
      }
    }
    else {
      console.warn( Config.errorPrefix + ' - ' + Config.errorNoConnection );
    }
  };

  self.net = net;

  self.placeOffer = function ( connectionId, connection, basket ) {

    var placeOfferCallback = function ( connectionId, error ) {
      var cId = connectionId.split( '@' );
      if ( error ) {
        console.log( cId[1] + ' - ' + Config.errorOffer );
        self.__finishHandler( connectionId, error );
        return;
      }
      else {
        console.log( cId[1] + ' - ' + Config.successOffer );
        self.__finishHandler( connectionId );
      }
    };

    var cId         = connectionId.split( '@' );
    var instance    = cId[1].trim();
    var isLive      = connection.events[basket.event].is_live;
    var isBuyerLive = connection.events[basket.event].is_buyer_live;
    var isActive    = connection.events[basket.event].is_active;
    var endpoint, payload;
    if ( !isLive && !isBuyerLive ) {
      payload  = {
        details  : ["user", connection.user_id, "pre_bid"],
        pre_bids : {}
      };
      payload.pre_bids[basket.id] = Math.floor( basket.offer / 100 * 100 ) / 100 ;
      endpoint = 'pre_bid';
    }
    else {
      payload  = {
        details : ["user", connection.user_id, "bid"],
        bids    : {}
      };
      payload.bids[basket.id] = Math.floor( basket.offer / 100 * 100 ) / 100 ;
      endpoint = 'bid';
    }

    console.log( Config.offerMsg );

    self.net.post( payload, 'events/user/' + connection.user_id + '/' + endpoint,
      function ( error ) {
        if ( !error )
          error = Config.networkError;
        connection.lastAction = Config.failedOfferMsg;
        console.error( Config.errorPrefix + ' ' + error );
        placeOfferCallback( connectionId, error );
      },
      function ( response ) {
        connection.lastAction    = Config.successOfferMsg;
        console.log( instance + ' - ' + Config.successOfferMsg );
        var payload              = JSON.parse( response.body ).payload;
        connection.lastActive    = new Date( payload.created );
        placeOfferCallback( connectionId );
      },
      instance,
      connection.sessionCookie
    );
  };

}
