// message-catalog/c2fo/invoice.js
var MCatCommon = require( 'message-catalog/common' ).MCatCommon;

exports.MCat = new MCat();

function MCat () {
  this.errorAmbiguous   = 'ambiguous instruction, skipping';
  this.errorInvoice     = 'could not retrieve invoices';
  this.errorToggle      = 'could not update invoices';
  this.failedInvoice    = 'could not sign in';
  this.failededToggle   = 'faileded to update invoices';
  this.invalidToken     = 'invalid token: ';
  this.invoiceCount     = 'total invoices retrieved: ';
  this.invoiceDetAction = 'retrieved invoices';
  this.invoiceDetail    = 'retrieved invoices';
  this.invoiceRetri     = 'starting invoice retrieval...';
  this.noEventFilter    = 'event filter cannot accompany date filter, skipping';
  this.noDateFilter     = 'date filter cannot accompany event filter, skipping';
  this.skipping         = ', skipping';
  this.successInvoice   = 'invoices retrieved';
  this.successToggle    = 'invoices updated';
  this.successToggle    = 'successful invoice update';
  this.toggle           = 'updating invoices...';

  for ( a in MCatCommon) {
    if ( a in this ) { continue; }
    this[a] = MCatCommon[a];
  }

}
