// config/event.js
exports.Config = new Config();

function Config () {
  this.awaitingStatus     = 'awaiting buyer close';
  this.BUYER              = 2;
  this.countdownMsg       = 'countdown to go-live';
  this.durationMsg        = 'duration';
  this.errorFlip          = 'could not flip supplier basket for: ';
  this.errorNoConnection  = 'no connection';
  this.errorOffer         = 'could not place offer';
  this.errorPrefix        = '[error]';
  this.errorSupplierCant  = "suppliers don't have this feature";
  this.eventsDetailAction = 'event check';
  this.eventsDetailMsg    = 'checked event schedule';
  this.failedOfferMsg     = 'failed to place offer';
  this.failedSignInMsg    = 'could not sign in';
  this.foundEventMsg      = 'found event #';
  this.liveStatus         = 'live';
  this.networkError       = 'check the login details, host or network';
  this.newCheckMsg        = 'new check date';
  this.offerMsg           = 'placing offer...';
  this.payThruMsg         = 'pay-thru date';
  this.preofferStatus     = 'accepting pre-offers';
  this.statusPrefix       = 'status';
  this.successFlip        = 'basket status flipped';
  this.successOffer       = 'offer placed';
  this.successOfferMsg    = 'successful offer';
  this.SUPPLIER           = 3;
  this.userDetailsAction  = 'read user details';
  this.userDetailsMsg     = 'retrieved user details';
}
