// config/invoice.js
exports.Config = new Config();

function Config () {
  this.BUYER              = 2;
  this.errorAmbiguous     = 'ambiguous instruction, skipping';
  this.errorInvoice       = 'could not retrieve invoices';
  this.errorNoConnection  = 'no connection';
  this.errorPrefix        = '[error]';
  this.errorToggle        = 'could not update invoices';
  this.eventsDetailAction = 'event check';
  this.eventsDetailMsg    = 'checked event schedule';
  this.failedSignInMsg    = 'could not sign in';
  this.failedToggleMsg    = 'failed to update invoices';
  this.invalidTokenMsg    = 'invalid token: ';
  this.invoiceCountMsg    = 'total invoices: ';
  this.invoiceRetriMsg    = 'starting invoice retrieval...';
  this.invoicesPageSize   = 40;
  this.networkError       = 'check the login details, host or network';
  this.noEventFilterMsg   = 'event filter cannot accompany date filter, skipping';
  this.noDateFilterMsg    = 'date filter cannot accompany event filter, skipping';
  this.skippingMsg        = ', skipping';
  this.successInvoice     = 'invoices retrieved';
  this.successToggle      = 'invoices updated';
  this.successToggleMsg   = 'successful invoice update';
  this.SUPPLIER           = 3;
  this.toggleMsg          = 'updating invoices...';
}
