// config/main.js
exports.Config = new Config();

function Config () {
  this.awaitingStatus     = 'awaiting buyer close';
  this.BUYER              = 2;
  this.connectMsg         = 'connecting...';
  this.countdownMsg       = 'countdown to go-live';
  this.durationMsg        = 'duration';
  this.errorFlip          = 'could not flip supplier basket for: ';
  this.errorInvoice       = 'could not retrieve invoices';
  this.errorNoConnection  = 'no connection';
  this.errorNoResponse    = 'no response data';
  this.errorOffer         = 'could not place offer';
  this.errorPrefix        = '[error]';
  this.errorSupplierCant  = "suppliers don't have this feature";
  this.errorToggle        = 'could not update invoices';
  this.eventsDetailAction = 'event check';
  this.eventsDetailMsg    = 'checked event schedule';
  this.failedOfferMsg     = 'failed to place offer';
  this.failedToggleMsg    = 'failed to update invoices';
  this.foundEventMsg      = 'found event #';
  this.invalidTokenMsg    = 'invalid token: ';
  this.invoiceCountMsg    = 'total invoices: ';
  this.invoicesPageSize   = 40;
  this.invoiceRetriMsg    = 'starting invoice retrieval...';
  this.liveStatus         = 'live';
  this.networkError       = 'check the login details, host or network';
  this.newCheckMsg        = 'new check date';
  this.noDateFilterMsg    = 'date filter cannot accompany event filter, skipping';
  this.noEventFilterMsg   = 'event filter cannot accompany date filter, skipping';
  this.offerMsg           = 'placing offer...';
  this.payThruMsg         = 'pay-thru date';
  this.preofferStatus     = 'accepting pre-offers';
  this.skippingMsg        = ', skipping';
  this.statusPrefix       = 'status';
  this.SUPPLIER           = 3;
  this.successOfferMsg    = 'successful offer';
  this.successToggleMsg   = 'successful invoice update';
  this.successFlip        = 'basket status flipped';
  this.successInvoice     = 'invoices retrieved';
  this.successOffer       = 'offer placed';
  this.successToggle      = 'invoices updated';
  this.toggleMsg          = 'updating invoices...';
  this.userDetailsAction  = 'read user details';
  this.userDetailsMsg     = 'retrieved user details';
}
