// message-catalog/common.js
var Util = require( 'util.js' ).Util;
exports.MCatCommon = new MCatCommon();

function MCatCommon () {
  this.BUYER              = 2;
  this.c                  = ': ';
  this.debugPrefix        = '[debug] ';
  this.e                  = '...';
  this.exit               = "leaving C2FO";
  this.failedJSONParse    = 'could not parse JSON';
  this.failedNoBuyer      = 'buyers cannot perform this action';
  this.failedNoConnection = 'no connection';
  this.failedPrefix       = '[error] ';
  this.help               = function ( f, s ) { var readme = Util.spitFile( 'help.txt', f, s ); };
  this.infoPrefix         = '[info] ';
  this.networkFail        = 'check the login details, host or network';
  this.notFound           = "command not found type 'h' for help";
  this.nullSymbol         = '*';
  this.promptPoint        = '> ';
  this.promptPrefix       = 'c2fo';
  this.s                  = ' - ';
  this.sn                 = ' - # ';
  this.SUPPLIER           = 3;
  this.warningPrefix      = '[warn] ';
}
