// message-catalog/ocap/event.js
var MCatCommon = require( 'message-catalog/common' ).MCatCommon;

exports.MCat = new MCat();

function MCat () {
  this.awaitingStatus     = 'awaiting buyer close';
  this.countdown          = 'countdown to go-live';
  this.deletingEvent      = 'deleting expired event #';
  this.duration           = 'duration';
  this.errorAward         = 'could not retrieve award data';
  this.errorFlip          = 'could not flip supplier basket for: ';
  this.errorOffer         = 'could not place offer';
  this.errorOffers        = 'could not retrieve offers';
  this.errorPreOffers     = 'could not retrieve pre-offers';
  this.errorSupplierCant  = "suppliers don't have this feature";
  this.eventsDetailAction = 'event check';
  this.eventsDetail       = 'checked event schedule';
  this.failededAward      = 'faileded to retrieve award data';
  this.failededEvent      = 'faileded to retrieve event details';
  this.failededLock       = 'faileded to lock basket';
  this.failededOffer      = 'faileded to place offer';
  this.failededOffers     = 'faileded to retrieve offer details';
  this.failededPreOffer   = 'faileded to retrieve pre-ffers';
  this.failededSignIn     = 'could not sign in';
  this.foundEvent         = 'found event #';
  this.liveStatus         = 'live';
  this.networkError       = 'check the login details, host or network';
  this.newCheck           = 'new check date';
  this.offer              = 'placing offer...';
  this.payThru            = 'pay-thru date';
  this.preofferStatus     = 'accepting pre-offers';
  this.statusPrefix       = 'status';
  this.successAward       = 'award data retrieved';
  this.successFlip        = 'basket status flipped';
  this.successOffer       = 'offer placed';
  this.successOffers      = 'offers retrieved';
  this.successOffer       = 'successful offer';
  this.successPreOffers   = 'pre-offers received';
  this.userDetailsAction  = 'read user details';
  this.userDetails        = 'retrieved user details';

  for ( a in MCatCommon) {
    if ( a in this ) { continue; }
    this[a] = MCatCommon[a];
  }

}
