// message-catalog/ocap/event.js
var MCatCommon = require( 'message-catalog/common' ).MCatCommon;

exports.MCat = new MCat();

function MCat () {
  this.awaitingStatus     = 'awaiting buyer close';
  this.countdownMsg       = 'countdown to go-live';
  this.deletingEventMsg   = 'deleting expired event #';
  this.durationMsg        = 'duration';
  this.errorFlip          = 'could not flip supplier basket for: ';
  this.errorOffer         = 'could not place offer';
  this.errorPreOffers     = 'could not retrieve pre-offers';
  this.errorSupplierCant  = "suppliers don't have this feature";
  this.eventsDetailAction = 'event check';
  this.eventsDetailMsg    = 'checked event schedule';
  this.failedOfferMsg     = 'failed to place offer';
  this.failedSignInMsg    = 'could not sign in';
  this.foundEventMsg      = 'found event #';
  this.liveStatus         = 'live';
  this.networkError       = 'check the login details, host or network';
  this.newCheckMsg        = 'new check date';
  this.offerMsg           = 'placing offer...';
  this.payThruMsg         = 'pay-thru date';
  this.preofferStatus     = 'accepting pre-offers';
  this.statusPrefix       = 'status';
  this.successFlip        = 'basket status flipped';
  this.successOffer       = 'offer placed';
  this.successOfferMsg    = 'successful offer';
  this.successPreOffers   = 'pre-offers received';
  this.userDetailsAction  = 'read user details';
  this.userDetailsMsg     = 'retrieved user details';

  for ( a in MCatCommon) {
    if ( a in this ) { continue; }
    this[a] = MCatCommon[a];
  }

}
