// message-catalog/ocap/invoice.js
var MCatCommon = require( 'message-catalog/common' ).MCatCommon;

exports.MCat = new MCat();

function MCat () {
  this.errorAmbiguous     = 'ambiguous instruction, skipping';
  this.errorInvoice       = 'could not retrieve invoices';
  this.errorToggle        = 'could not update invoices';
  this.failedSignInMsg    = 'could not sign in';
  this.failedToggleMsg    = 'failed to update invoices';
  this.invalidTokenMsg    = 'invalid token: ';
  this.invoiceCountMsg    = 'total invoices: ';
  this.invoiceDetAction   = 'retrieved invoices';
  this.invoiceDetailMsg   = 'retrieved invoices';
  this.invoiceRetriMsg    = 'starting invoice retrieval...';
  this.invoicesPageSize   = 40;
  this.networkError       = 'check the login details, host or network';
  this.noEventFilterMsg   = 'event filter cannot accompany date filter, skipping';
  this.noDateFilterMsg    = 'date filter cannot accompany event filter, skipping';
  this.skippingMsg        = ', skipping';
  this.successInvoice     = 'invoices retrieved';
  this.successToggle      = 'invoices updated';
  this.successToggleMsg   = 'successful invoice update';
  this.toggleMsg          = 'updating invoices...';

  for ( a in MCatCommon) {
    if ( a in this ) { continue; }
    this[a] = MCatCommon[a];
  }

}
